rfcs/0015-build-phases.md

```diff
@@ -0,0 +1,130 @@
+
+### Additional Goal
+*(to be investigated)*
+
+This concept might make it possible for tasks and project builds to run semi-concurrently. For instance, certain build phases could be declared as "dependency free", meaning tasks in those phases must only process resources of the project itself and do not depend on the build result of dependencies. This would allow UI5 Tooling to executed those tasks concurrently while other projects are still being built.
+
+Similarly, we could define that tasks in the same build phase must not depend on the result of each other, allowing for executing them concurrently as well.
+
+The actual performance improvements from such changes should be proven first however.
+
+## Detailed design
+
+### Build Phases
+
+Build Phase | Standard Tasks in that Phase | Notes
+----: | ---- | ----
+| | | |
+_pre-build_ | *none* |
+| | | |
+_pre-prepare-sources_ | *none* | *No dependency access?*
+**prepare-sources** | escapeNonAsciiCharacters, replaceCopyright, replaceVersion, replaceBuildtime, minify | *No dependency access?*
+_post-prepare-sources_ | *none* | *No dependency access?*
+| | | |
+_pre-create-bundles_ | *none* |
+**create-bundles** | generateFlexChangesBundle, generateManifestBundle,<br/>generateLibraryManifest, generateComponentPreload,<br/>generateLibraryPreload, generateStandaloneAppBundle |
+_post-create-bundles_ | transformBootstrapHtml |
+| | | |
+_pre-build-themes_ | *none* |
+**build-themes** | buildThemes |
+_post-build-themes_ | generateThemeDesignerResources |
+| | | |
+_post-build_ | generateVersionInfo, generateVersionInfo, generateApiIndex, generateResourcesJson | Or rather a root-project specific phase?
+
+The custom task configuration can reference a build phase and define whether the task should be executed at the *start* or at the *end* of a given phase. If multiple custom tasks are defined for the same execution time, the order of their definition in the project's ui5.yaml configuration is relevant.
+
+**TODO:** Allow for concurrent execution in some phases?`
+
+### Custom Task Configuration
+
+**Old**
+
+```yaml
+specVersion: "3.0"
+type: library
+metadata:
+  name: my.library
+builder:
+  customTasks:
+    - name: render-markdown-files
+      afterTask: minify # <- Explicit dependency to minify task
+---
+specVersion: "3.0"
+kind: extension
+type: task
+metadata:
+  name: render-markdown-files
+task:
+  path: lib/tasks/renderMarkdownFiles.js
+```
+
+**New**
+
+```yaml
+specVersion: "4.0"
+type: library
+metadata:
+  name: my.library
+builder:
+  customTasks:
+    - name: render-markdown-files
+      # No scheduling configuration necessary (although possible, overwriting the task's default)
+---
+specVersion: "4.0"
+kind: extension
+type: task
+metadata:
+  name: render-markdown-files
+task:
+  path: lib/tasks/renderMarkdownFiles.js
+  defaultBuildPhase: post-create-bundles # <- New default configuration
```

Comment:
IMO, the `default` part is a bit missleading. Maybe, skipping it would make things clearer.
`defaultBuildPhase` seems like it implies changing the default phase for everything which is not the case here. Skipping `default` from the name makes it clear that this task will be executed at that phase

```suggestion
  buildPhase: post-create-bundles # <- New default configuration
```